server/src/internalClusterTest/java/org/elasticsearch/action/support/master/TransportMasterNodeActionIT.java

```diff
@@ -36,39 +39,49 @@ protected Collection<Class<? extends Plugin>> nodePlugins() {
         return CollectionUtils.appendToCopy(super.nodePlugins(), MockTransportService.TestPlugin.class);
     }
 
-    @TestLogging(reason = "wip", value = "org.elasticsearch.transport.TransportService.tracer:TRACE")
+    @TestLogging(reason = "wip", value = "org.elasticsearch.action.support.master.TransportMasterNodeAction:DEBUG")
     public void testRoutingLoopProtection() {
 
         final var newMaster = internalCluster().startMasterOnlyNode();
         final var enoughVotingMastersLatch = new CountDownLatch(1);
-        final var electionWonLatch = new CountDownLatch(1);
-        final var releaseMasterLatch = new CountDownLatch(1);
-
-        final ClusterStateApplier blockingApplier = event -> {
-            if (3 <= event.state().coordinationMetadata().getLastCommittedConfiguration().getNodeIds().size()) {
-                enoughVotingMastersLatch.countDown();
-            }
-
-            if (event.state().nodes().isLocalNodeElectedMaster()) {
-                logger.info("--> new master elected as planned");
-                electionWonLatch.countDown();
-                safeAwait(releaseMasterLatch);
-                logger.info("--> cluster state applications released on new master");
-            }
-        };
+        final var newMasterReceivedReroutedMessageLatch = new CountDownLatch(1);
 
         try {
-            internalCluster().getInstance(ClusterService.class, newMaster).addStateApplier(blockingApplier);
-
-            // need at least 3 voting master nodes for failover
+            /*
+             * Ensure that we've got 5 voting nodes in the cluster, this means even if the original
+             * master manages to accept its own failed state update before standing down, we can still
+             * establish a quorum without its (or our own) join.
+             */
+            internalCluster().getInstance(ClusterService.class, newMaster).addStateApplier(event -> {
+                if (5 <= event.state().coordinationMetadata().getLastCommittedConfiguration().getNodeIds().size()) {
+                    enoughVotingMastersLatch.countDown();
+                }
+            });
+            internalCluster().startMasterOnlyNode();
+            internalCluster().startMasterOnlyNode();
             internalCluster().startMasterOnlyNode();
             safeAwait(enoughVotingMastersLatch);
```

Comment:
This feels a bit hacky, but I struggled to see how we could
1. block the applier on newMaster to prevent it clearing the current master
2. allow newMaster to send a joinRequest to itself
3. block the applier on newMaster to prevent it applying the cluster state after it wins an election

I think (although with all the asynchrony, I'm not sure) that the above things happen in the above order, so I don't know how we could block 1 while allowing 2 to go through.

That means we can't rely on our own join, and because the old master sometimes accepts its failed update, we can't rely on its vote, so we can't reliably get a quorum in a 3 node cluster. If we bump the cluster size up to 5 nodes we seem to succeed every time.